### PR TITLE
dec 20 tele fix

### PR DESCRIPTION
**Pull request by valaises**

### CODE REVIEW

File: src/telemetry/basic_transmit.rs [hunk 1]
```diff
@@ -141,9 +146,15 @@ pub async fn telemetry_full_cycle(
...
+    let mut skip_compression_part = true;
     loop {
         tokio::time::sleep(tokio::time::Duration::from_secs(TELEMETRY_TRANSMIT_AFTER_START_SECONDS)).await;
-        telemetry_full_cycle(global_context.clone(), false).await;
+        telemetry_full_cycle(
+            global_context.clone(),
+            false,
+            skip_compression_part,
+        ).await;
+        skip_compression_part = false;
```
Comment (olegklimov) [inline src/telemetry/basic_transmit.rs:157]: bad code

File: src/telemetry/basic_transmit.rs [hunk 2]
```diff
@@ -111,9 +112,13 @@ pub async fn telemetry_full_cycle(
...
-    basic_network::compress_basic_telemetry_to_file(global_context.clone()).await;
-    basic_robot_human::tele_robot_human_compress_to_file(global_context.clone()).await;
-    basic_comp_counters::compress_tele_completion_to_file(global_context.clone()).await;
+    if !skip_compression_part {
+        basic_network::compress_basic_telemetry_to_file(global_context.clone()).await;
+        basic_robot_human::tele_robot_human_compress_to_file(global_context.clone()).await;
+        basic_comp_counters::compress_tele_completion_to_file(global_context.clone()).await;
+    } else {
+        info!("skip_compression_part is true, skip");
+    }
```
Comment (olegklimov) [inline src/telemetry/basic_transmit.rs:121]: we can split this function in two, delete flags

File: src/telemetry/basic_transmit.rs [hunk 3]
```diff
@@ -17,7 +17,7 @@ use crate::telemetry::utils::{sorted_json_files, read_file, cleanup_old_files, t
 
 const TELEMETRY_TRANSMIT_AFTER_START_SECONDS: u64 = 60;
 const TELEMETRY_TRANSMIT_EACH_N_SECONDS: u64 = 3600*3;
-const TELEMETRY_FILES_KEEP: i32 = 30;
+const TELEMETRY_FILES_KEEP: i32 = 128;
```
Comment (olegklimov) [inline src/telemetry/basic_transmit.rs:21]: debug code?

Comment (valaises) [reply src/telemetry/basic_transmit.rs:21]: no. files are small enough to keep more files. I'm afraid we could loose some files with limit=30

File: src/telemetry/basic_transmit.rs [hunk 4]
```diff
@@ -141,9 +142,18 @@ pub async fn telemetry_full_cycle(
...
-        telemetry_full_cycle(global_context.clone(), false).await;
+
+        if !skip_compression_part {
+            basic_telemetry_compress(global_context.clone()).await;
+        }
+        skip_compression_part = false;
+
+        basic_telemetry_send(global_context.clone()).await;
+
+
```
Comment (olegklimov) [inline src/telemetry/basic_transmit.rs:156]: can be better

File: src/telemetry/utils.rs [hunk 5]
```diff
@@ -19,24 +19,57 @@ pub async fn telemetry_storage_dirs(cache_dir: &PathBuf) -> (PathBuf, PathBuf) {
...
             ChangeTag::Delete => {
+                // if removed.is_empty()  {
+                //     removed += change.value();
+                // }
+                // info!("rem: {}; len: {}", change.value(), change.value().len());
+                if only_one_deletion_allowed {
+                    if !removed.is_empty() {
+                        continue;
+                    }
+                }
```
Comment (olegklimov) [inline src/telemetry/utils.rs:61]: counts only the first

File: src/telemetry/telemetry_structs.rs [hunk 6]
```diff
@@ -23,10 +25,19 @@ impl Storage {
...
     }
 }
 
+
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+pub struct InitFileText {
+    pub uri: String,
+    pub file_text: String,
+}
+
```
Comment (olegklimov) [inline src/telemetry/telemetry_structs.rs:40]: no

File: src/telemetry/basic_robot_human.rs [hunk 7]
```diff
@@ -15,76 +15,83 @@ use crate::telemetry::telemetry_structs;
...
 
     if let Some(rec) = tele_robot_human.iter_mut().find(|stat| stat.uri.eq(uri)) {
         if rec.used_snip_ids.contains(&snip.snippet_telemetry_id) {
             return;
         }
-        let robot_characters = robot_characters(snip);
-        rec.robot_characters_acc_baseline += robot_characters;
+
+        update_robot_characters_baseline(rec, snip);
+        update_human_characters(rec, rec.baseline_text.clone(), text);
```
Comment (olegklimov) [inline src/telemetry/basic_robot_human.rs:58]: rename

File: src/telemetry/basic_robot_human.rs [hunk 8]
```diff
@@ -15,76 +15,83 @@ use crate::telemetry::telemetry_structs;
...
+) {
+    let re = Regex::new(r"\s+").unwrap();
+    let (added_characters, removed_characters) = utils::get_add_del_from_texts(&baseline_text, text, true);
+    let real_characters_added = re.replace_all(&added_characters, "").len() as i64 - re.replace_all(&removed_characters, "").len() as i64;
+    let human_characters = real_characters_added - rec.robot_characters_acc_baseline;
+    info!("human_characters: +{}; robot_characters: +{}", human_characters, rec.robot_characters_acc_baseline);
+    rec.human_characters += human_characters;
+    rec.robot_characters += rec.robot_characters_acc_baseline;
+    rec.robot_characters_acc_baseline = 0;
+}
```
Comment (olegklimov) [inline src/telemetry/basic_robot_human.rs:40]: basetext_to_text_leap_calculations

Comment (olegklimov) [reply src/telemetry/basic_robot_human.rs:40]: added_lines, removed_lines

Comment (olegklimov) [reply src/telemetry/basic_robot_human.rs:40]: added_chars, removed_chars = diff(removed_lines, added_lines)
for line in added_lines {
   added_chars_this_line = ...;
   info!("line \"{}\" < human added characters \"{}\", added_chars_this_line);
}

File: src/telemetry/utils.rs [hunk 9]
```diff
@@ -19,28 +19,78 @@ pub async fn telemetry_storage_dirs(cache_dir: &PathBuf) -> (PathBuf, PathBuf) {
...
     text_b: &String,
+    only_one_deletion_allowed: bool,
 ) -> (String, String) {
-    let diff = TextDiff::from_lines(text_a, text_b);
+    let mut text_a_lines = text_a.lines().collect::<Vec<&str>>();
+    let mut text_b_lines = text_b.lines().collect::<Vec<&str>>();
+
+    for s in &mut text_a_lines {
+        *s = s.trim_end().trim_start();
+        // info!("text_a: {}; len: {}", s, s.len());
```
Comment (reymondzzzz) [inline src/telemetry/utils.rs:29]: if u not sure, use debug!() macros. I will add extra param for debug lvl https://docs.rs/tracing/latest/tracing/macro.debug.html

File: src/telemetry/utils.rs [hunk 10]
```diff
@@ -19,28 +19,78 @@ pub async fn telemetry_storage_dirs(cache_dir: &PathBuf) -> (PathBuf, PathBuf) {
...
+    }
+
+    for s in &mut text_b_lines {
+        *s = s.trim_end().trim_start();
+        // info!("text_b: {}; len: {}", s, s.len());
```
Comment (reymondzzzz) [inline src/telemetry/utils.rs:34]: same

File: src/telemetry/basic_robot_human.rs [hunk 11]
```diff
@@ -15,7 +15,51 @@ use crate::telemetry::telemetry_structs;
...
+    rec: &mut TeleRobotHumanAccum,
+    baseline_text: String,
+    text: &String,
+) {
+    let re = Regex::new(r"\s+").unwrap();
+    let (added_characters, removed_characters) = utils::get_add_del_from_texts(&baseline_text, text, false);
+
+    let (added_characters, _) = utils::get_add_del_chars_from_texts(&removed_characters, &added_characters);
+
+    // let real_characters_added = re.replace_all(&added_characters, "").len() as i64 - re.replace_all(&removed_characters, "").len() as i64;
```
Comment (reymondzzzz) [inline src/telemetry/basic_robot_human.rs:56]: remove extra comments

File: src/telemetry/basic_robot_human.rs [hunk 12]
```diff
@@ -24,67 +68,40 @@ pub fn increase_counters_from_finished_snippet(
...
 
-    for accum in data {
+    for accum in tele_robot_human {
         let key = (accum.file_extension.clone(), accum.model.clone());
         unique_combinations.entry(key).or_default().push(accum);
     }
     let mut compressed_vec= vec![];
     for (key, entries) in unique_combinations {
-        info!("compress_robot_human: compressing {} entries for key {:?}", entries.len(), key);
+        // info!("compress_robot_human: compressing {} entries for key {:?}", entries.len(), key);
```
Comment (reymondzzzz) [inline src/telemetry/basic_robot_human.rs:104]: same with debug!

File: src/telemetry/basic_transmit.rs [hunk 13]
```diff
@@ -141,9 +142,11 @@ pub async fn telemetry_full_cycle(
...
 ) -> () {
+    tokio::time::sleep(tokio::time::Duration::from_secs(TELEMETRY_TRANSMIT_AFTER_START_SECONDS)).await;
+    basic_telemetry_send(global_context.clone()).await;
     loop {
-        tokio::time::sleep(tokio::time::Duration::from_secs(TELEMETRY_TRANSMIT_AFTER_START_SECONDS)).await;
-        telemetry_full_cycle(global_context.clone(), false).await;
-        tokio::time::sleep(tokio::time::Duration::from_secs(TELEMETRY_TRANSMIT_EACH_N_SECONDS - TELEMETRY_TRANSMIT_AFTER_START_SECONDS)).await;
+        tokio::time::sleep(tokio::time::Duration::from_secs(TELEMETRY_TRANSMIT_EACH_N_SECONDS)).await;
+        basic_telemetry_send(global_context.clone()).await;
+        basic_telemetry_compress(global_context.clone()).await;
```
Comment (olegklimov) [inline src/telemetry/basic_transmit.rs:150]: swap

**Pull request merged**